### PR TITLE
Publish sequence readme on website

### LINKED ISSUE #6264: Sequence package readme should be published on fluidframework.com
The [sequence readme](https://github.com/microsoft/FluidFramework/blob/main/packages/dds/sequence/README.md) should be published on the website so it's easier to find.

### PR DESCRIPTION
**Pull request by tylerbutler**
Fixes #6264.

See it in staging: 
https://agreeable-hill-0eeff5b10-6265.centralus.azurestaticapps.net/docs/data-structures/sequence/

### CODE REVIEW

File: packages/dds/sequence/README.md [hunk 1]
```diff
@@ -131,12 +168,23 @@ Both markers and text are stored as segments in the Shared String. Text segments
   - [collaborativeTextArea](https://github.com/microsoft/FluidFramework/blob/main/examples/data-objects/react-inputs/src/CollaborativeTextArea.tsx)
   - [collaborativeInput](https://github.com/microsoft/FluidFramework/blob/main/examples/data-objects/react-inputs/src/collaborativeInput.tsx)
 
-# Sparse Matrix
+## SparseMatrix (obsolete)
+
+**OBSOLETE:** SparseMatrix has been replaced by SharedMatrix.
```
Comment (tylerbutler) [inline packages/dds/sequence/README.md:173]: @anthony-murphy @DLehenbauer Is SparseMatrix obsolete?

File: docs/content/docs/data-structures/sequence.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+differs from an array in that the positions can move as local and remote collaborators make modifications to the
+sequence. There are a number of different sequence types:
+
+- SharedString for storing and collaborating on a sequence of text.
+- SharedNumberSequence for storing and collaborating on a sequence of numbers.
+- SharedObjectSequence for storing and collaborating on a sequence of JSON-serializable objects.
+
+As the name suggests SharedSegmentSequence, or sequence for short, are made of segments. Segments are the leaf nodes of
+the tree data structure that enables collaboration and backs the sequence. Segments may be split and merged as
+modifications are made to the sequence. Every segment has a length from 1, to the length of the sequence. The length of
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md:21]: "Every segment has a length from 1", or "every segment is counted to be an single element, and length of sequence is number of elements"? The former suggests that it can be not 1 and needs more explanation.

File: packages/dds/sequence/README.md (hunk 1 above)
Comment (DLehenbauer) [reply packages/dds/sequence/README.md:173]: Correct.  I'd consider omitting SparseMatrix from the README.  It's replacement (SharedMatrix) is better in every way.

Comment (tylerbutler) [reply packages/dds/sequence/README.md:173]: I'll remove the docs from the readme and filed a follow up to mark the code itself as obsolete in the TSDocs.

File: docs/content/docs/data-structures/sequence.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+the sequence will be the combined length of all the segments.
+
+When talking about positions in a sequence we use the terms *near*, and *far*. The nearest position in a sequence is 0,
+and the farthest position is its length. When comparing two positions the nearer position is closer to 0, and the
+farther position is closer to the length.
+
+## Using a Sequence
+
+Sequences support three basic operations: insert, remove, and annotate.
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md:30]: There was no mention to properties, so "annotate" here may feel unfamiliar when it is introduced here.

File: docs/content/docs/data-structures/sequence.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+    // insert text at position 2
+    sharedString.insertText(
+        2,
+        " world");
+    //   content: hi world!
+    // positions: 012345678
+```
+
+Remove operations take a start and an end position. The start position is similar to the insert's position, in that is
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md:59]: I feel like there is disconnect in the flow RE segments and position. I.e. till now we said that data is represented as segments and segment length is 1. We did not introduce how test is mapped to segments, so uneducated user may assume that position in this context maps to directly to segments, i.e. end-start+1 is number of segments impacted by this operation, which is not the case.

Comment (vladsud) [reply docs/content/docs/data-structures/sequence.md:59]: Also probably worth clarifying that sharedString.getLength() is not number of segments, it's number of characters.
Somewhere we need to say that if you enumerate segments, some of them will be "deleted" segments.

File: docs/content/docs/data-structures/sequence.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    // remove all the characters
+    sharedString.removeRange(0, sharedString.getLength());
+    //   content:
+    // positions:
+```
+
+Annotate operations can add or remove map-like properties to or from content of the sequence. They can store any JSON
+serializable data and have similar behavior to a shared map. Annotate takes a start and end position which work the same
+way as the start and end of the remove operation. In addition to start and end annotate also takes a map-like properties
+object. Each key of the provided properties object will be set on each position of the specified range. Setting a
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md:82]: You probably want to say here how properties are merged or overwritten (and mentioned what is the default mode)

File: docs/content/docs/data-structures/sequence.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    // props5 = { decoration: "underline" }
+```
+
+Whenever an operation is performed on a sequence a *sequenceDelta* event will be raised. This even provides the ranges
+affected by the operation, the type of the operation, and the properties that were changes by the operation.
+
+## How Collaboration Works
+
+Like other data structures the sequences are eventually consistent which means all collaborators will end up in the same
+final state, however, the intermediate states seen by each collaborator may not be seen by other collaborators. These
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md:128]: To be pedantic - intermediate states happen even if there is no conflict

File: docs/content/docs/data-structures/sequence.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+intermediate states occur when two or more collaborators modify the same position in the sequence which results in a
+conflict.
+
+The basic strategy for insert conflict resolution in the sequence is to merge *far*. This strategy depends on a
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md:132]: I find this paragraph too complicated for describing process, and using terminology (far) that is not helping (and not used anywhere else).

I think you can simplify it by basically saying that conflicting inserts are resolved as if there was no conflict and changes were spaced out in time, with change hitting server earlier (user 1) being observed by second client but not moving his cursor (insertion point), thus making his/her (later) insert go to the left of first user's insert.

That said, does it really matter? It seems like it's implementation detail that should not leak out, unless someone builds some rather complicated layer on top of it (which is advanced topic).

File: docs/content/docs/data-structures/sequence.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+fundamental property of the Fluid Framework, which is guaranteed ordering. So, if two or more collaborators perform an
+operation on a sequence, the operations will be given an ordering and all clients will see those operations in the same
+order. What this means for the merge *far* strategy for resolving conflicting inserts is that the first operation will
+be placed in the conflicting position when it is received. When the next insert with the same position arrives and is
+applied it will be placed at the specified position and the previous inserts content position will be increased by the
+length of the incoming content pushing is farther towards the length of the sequence. This is what we call merging
+*far*.
+
+Like insert the strategies for remove and annotate also rely on guaranteed ordering. For remove and annotate only
+content visible to the collaborator creating the operation will be modified, any content ordered after the won't be.
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md:143]: I do not think this paragraph answer the question of what happens if user A inserts at pos = 2, while user B deletes 0..4 range. Or at least - most people learn by samples, and it would be great to have that sample right after that paragraph.

File: docs/content/docs/data-structures/sequence.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+For remove this means we can't have an insert and a remove at the same time, as they will have an order, and all
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md:145]: > For remove this means we can't have an insert and a remove at the same time, as they will have an order, and all collaborators will see the operations in the same order

I'm not sure what exactly this sentence tries to say :)

File: docs/content/docs/data-structures/sequence.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+collaborators will see the operations in the same order. We also detect overlapping removes made by different
+collaborators, the resolutions here is straightforward, the content is removed.
+
+As mentioned above annotate operations behave like operations on Shared Maps. The merge strategy here is last one wins.
+So, if two collaborators set the same key on the annotates properties the operation that gets ordered last will
```
Comment (vladsud) [inline docs/content/docs/data-structures/sequence.md]: > The merge strategy here is last one wins

I believe this is default strategy, but not the only one.

File: docs/content/docs/data-structures/sequence.md (hunk 2 above)
Comment (DLehenbauer) [reply docs/content/docs/data-structures/sequence.md:21]: A SharedSequence is an array-like ordered sequence of items.  SharedSequence operations are defined in terms of the item's position in the sequence (e.g., removeRange(start, end)).

For efficiency, a SharedSequence stores contiguous blocks of positions/items in segments.  This is mostly an under-the-covers detail.  However, we do expose the ability to enumerate the segments that intersect a range of positions.  Enumerating the segments allows a savvy renderer to avoid per-character tree lookups and string allocations as it builds up the HTML text content.

The intent of the above prose is to communicate that:

a) Segments contain 1 or more positions (i.e., have a minimum length of 1).
b) Segments may be arbitrarily split/merged by the SharedSequence.  Users should generally not retain references to segments outside of the enumeration or have expectations about segment length.

File: docs/content/docs/data-structures/sequence.md (hunk 9 above)
Comment (tylerbutler) [reply docs/content/docs/data-structures/sequence.md:145]: This is trying to say that the concept of "at the same time" doesn't exist in Fluid since ops are explicitly ordered.

File: docs/content/docs/data-structures/sequence.md (hunk 7 above)
Comment (tylerbutler) [reply docs/content/docs/data-structures/sequence.md:133]: > I find right and left terminology is much more on point to describing what actually happens, and I've heard our own team using it (never heard "far" terminology).

I think one reason the near/far terminology was used is because for RTL languages, left/right isn't accurate. That was how @anthony-murphy explained it way back when. That said, I don't have a strong attachment to near/far.

I don't think any amount of words will be as helpful as some diagrams or animations of the algorithm in action if you **really** want to know what's going on under the covers. I dream of someday having those. 😄 But I think you're right - this detail shouldn't be relevant to general usage. I do love having that detail available for those who dig, but maybe that detail belongs in the contributor docs (currently the wiki)? Or we could add a "how it works" section, which has come up before. I'll mull it over.

File: docs/content/docs/data-structures/sequence.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+For remove this means we can't have an insert and a remove at the same time, as they will have an order, and all
+collaborators will see the operations in the same order. We also detect overlapping removes made by different
+collaborators, the resolutions here is straightforward -- the content is removed.
+
+As mentioned above, annotate operations behave like operations on SharedMaps. The merge strategy here is last one wins.
+So, if two collaborators set the same key on the annotate properties the operation that gets ordered last will
+determine the value.
+
+## SharedString
```
Comment (Rick-Kirkham) [inline docs/content/docs/data-structures/sequence.md:153]: Many of the code examples above used sharedString, so isn't most of this section redundant? Maybe what we need is only a section on the ways that SharedStrings differ from other sequences. E.g., the markers stuff.

File: docs/content/docs/data-structures/sequence.md [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+## SharedObjectSequence and SharedNumberSequence
+
+SharedObjectSequence and SharedNumberSequence are very similar distributed data structures. The only difference is the
+type of content they support. SharedNumberSequence only supports numbers as content, while SharedObjectSequence supports
+any JSON serializable object. Both DDSes support inserting, removing, and annotating content. Each piece of content --
+that is, each number or object -- will occupy a single position in the sequence. The length of the sequence is the count
+of content items in the sequence.
+
+An important note is that, unlike an array, positions are not guaranteed remain constant. The position of an item can
```
Comment (Rick-Kirkham) [inline docs/content/docs/data-structures/sequence.md:193]:
```suggestion
An important note is that, unlike an array, positions are not guaranteed to remain constant. The position of an item can
```

File: docs/content/docs/data-structures/sequence.md [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+change as content is added or removed from the sequence. To track or pass a reference to a specific piece of content
```
Comment (Rick-Kirkham) [inline docs/content/docs/data-structures/sequence.md:194]: This looks like a really important technique that should get more emphasis and a concrete example.

File: packages/dds/sequence/README.md [hunk 14]
```diff
@@ -1,19 +1,29 @@
...
+- SharedString for storing and collaborating on a sequence of text.
+- SharedNumberSequence for storing and collaborating on a sequence of numbers.
+- SharedObjectSequence for storing and collaborating on a sequence of JSON-serializable objects.
 
-When talking about positions in a sequence we use the terms _near_, and _far_. The nearest position in a sequence is 0, and the farthest position is its length. When comparing two positions the nearer position is closer to 0, and the farther position is closer to the length.
+As the name suggests SharedSegmentSequence, or sequence for short, are made of segments. Segments are the leaf nodes of
+the tree data structure that enables collaboration and backs the sequence. Segments may be split and merged as
+modifications are made to the sequence. Every segment has a length from 1, to the length of the sequence. The length of
```
Comment (Rick-Kirkham) [inline packages/dds/sequence/README.md:13]: We're using "length of the sequence" in many places the way a variable is used in code. But it takes the reader a while to figure this out. As an alterative, I suggest using "n (where n is the length of the sequence)". In this case:

```suggestion
modifications are made to the sequence. Every segment has a length from 1 to *n* (where *n* is the length of the sequence). The length of
```
An alternative, but not as good would be to hyphenate so readers understand that it is really one word: "from 1 to length-of-the-sequence".

**Review by Rick-Kirkham — CHANGES_REQUESTED** [review]
Most important thing: Add a lot of concrete examples, early and often!

File: packages/dds/sequence/README.md (hunk 1 above)
Comment (Dhoot) [reply packages/dds/sequence/README.md]: but it is present in code [packages/dds/sequence/src/sparsematrix.ts](https://github.com/microsoft/FluidFramework/blob/main/packages/dds/sequence/src/sparsematrix.ts) so do need to remove it everywhere or only from docs ?

File: docs/content/docs/data-structures/sequence.md (hunk 2 above)
Comment (vladsud) [reply docs/content/docs/data-structures/sequence.md:21]: I'd rephrase it. I'd say every segment has some length, and sum of segment lengths is the length of sequence.

File: docs/content/docs/data-structures/sequence.md (hunk 7 above)
Comment (vladsud) [reply docs/content/docs/data-structures/sequence.md:133]: I do not think Sequence cares about language - resolution always happens in exactly same way no matter what is inserted (text or other types of segments like markers). And I think the meaning of "right" & "left" words is not changed in RTL languages.

File: docs/content/docs/data-structures/sequence.md (hunk 9 above)
Comment (vladsud) [reply docs/content/docs/data-structures/sequence.md:145]: I do not think Fluid brings anything unique here.
Most processors (Intel x86 family including) have global ordering guarantee, i.e. all parties (threads) observe events in same order. I.e. if one party (thread) saw A happen after B, then all other parties/threads will see these events in exactly same order.
But in multi-threading, we use "at the same time" terminology (or rather more correct - "concurrent" changes), despite the fact that there will be order assigned to those events.

If we want to not use "at the same time" terminology, then we need to come up with some other terminology that describes changes that happened so that each of it had no chance to observe the other at the time change was made (aka concurrent changes).

**Comment by tylerbutler** [comment]
@Rick-Kirkham @vladsud @anthony-murphy @DLehenbauer I've made a lot of changes to this doc in an attempt to clarify it. I'd like to merge this in by the end of the month. I don't think there's anything outright wrong now, and we can continue to iterate,

Note that the small section on SharedString is only in the README. We'll have a separate doc for SharedString on fluidframework,com.

**Review by Rick-Kirkham — APPROVED** [review]
